core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java

```diff
@@ -139,82 +76,54 @@ private boolean waitForChannelReady(ManagedChannel managedChannel, long healthCh
    * @param shutdownTimeoutMs shutdown timeout in milliseconds
    * @return a {@link ManagedChannel}
    */
-  public ManagedChannel acquireManagedChannel(GrpcChannelKey channelKey,
-      long healthCheckTimeoutMs, long shutdownTimeoutMs) {
-    boolean shutdownExistingChannel = false;
-    ManagedChannelReference managedChannelRef = null;
-    try (LockResource lockShared = new LockResource(mLock.readLock())) {
-      if (mChannels.containsKey(channelKey)) {
-        managedChannelRef = mChannels.get(channelKey);
-        if (waitForChannelReady(managedChannelRef.get(),
-            healthCheckTimeoutMs)) {
-          LOG.debug("Acquiring an existing managed channel. ChannelKey: {}. Ref-count: {}",
-              channelKey, managedChannelRef.getRefCount());
-          return managedChannelRef.reference();
+  public ManagedChannel acquireManagedChannel(GrpcChannelKey channelKey, long healthCheckTimeoutMs,
+      long shutdownTimeoutMs) {
+    return mChannels.compute(channelKey, (key, chHolder) -> {
+      boolean shutdownExistingChannel = false;
+      int existingRefCount = 0;
+      if (chHolder != null) {
+        if (waitForChannelReady(chHolder.get(), healthCheckTimeoutMs)) {
+          LOG.debug("Acquiring an existing managed channel. ChannelKey: {}. Ref-count: {}", key,
+              chHolder.getRefCount());
+          return chHolder.reference();
         } else {
-          // Postpone channel shutdown under exclusive lock below.
           shutdownExistingChannel = true;
         }
       }
-    }
-    try (LockResource lockExclusive = new LockResource(mLock.writeLock())) {
-      // Dispose existing channel if required.
-      int existingRefCount = 0;
-      if (shutdownExistingChannel && mChannels.containsKey(channelKey)
-          && mChannels.get(channelKey) == managedChannelRef) {
-        existingRefCount = managedChannelRef.getRefCount();
+      if (shutdownExistingChannel) {
+        existingRefCount = chHolder.getRefCount();
```

Comment:
Why do we preserve the old ref count if we are shutting down the channel? I don't think the other users of the channel be able to pick up the new healthy channel we create?

Reply:
Due to shutdown, their operations should start failing and eventually they'll release their channels. When they reacquire after release, that's when they'll get the new healthy channel.

Reply:
So setting the old ref count is to account for the future `release` calls with the same channelKey on the now shutdown channel?

Reply:
Would it be better to pass in the managed channel in the release method so the release of an already shutdown channel becomes a noop instead of affecting the reference counting of the new channel?

Reply:
@calvinjia Correct.
@bf8086 That will require isolating the resource from the pool. I'm not sure what we'll gain from doing that.

Reply:
All the callers of the release method has access to the resource(managed channel) already so they can just pass in. It feels more clear to me when we can explicitly specify which actual managed channel we release so we don't have to worry about dereference new channel from users of the old channel.

Reply:
I agree this is not very clean. However I don't feel good about requiring both the key and the channel itself during `release`. Looks like it could be done clearly if we return a pool-handle from `acquire` call, rather than returning a channel.. Put a TODO for addressing this later.

Reply:
https://github.com/Alluxio/alluxio/issues/9831